cmd/notation/manifest.go

```diff
@@ -3,34 +3,119 @@ package main
 import (
 	"context"
 	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode"
 
 	"github.com/notaryproject/notation-go/log"
 	notationregistry "github.com/notaryproject/notation-go/registry"
+	notationerrors "github.com/notaryproject/notation/cmd/notation/internal/errors"
+	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"oras.land/oras-go/v2/registry"
 )
 
-// getManifestDescriptor returns target artifact manifest descriptor and
-// registry.Reference given user input reference.
-func getManifestDescriptor(ctx context.Context, opts *SecureFlagOpts, reference string, sigRepo notationregistry.Repository) (ocispec.Descriptor, registry.Reference, error) {
-	logger := log.GetLogger(ctx)
+const ociLayout = "oci-layout"
 
+// resolveReference resolves user input reference based on user input type.
+// Returns the resolved manifest descriptor, a full representation of
+// the reference in digest, and a print out for reference in digest
+func resolveReference(ctx context.Context, inputType inputType, reference, policyScope string, sigRepo notationregistry.Repository, fn func(string, ocispec.Descriptor)) (ocispec.Descriptor, string, string, error) {
+	// sanity check
 	if reference == "" {
-		return ocispec.Descriptor{}, registry.Reference{}, errors.New("missing reference")
+		return ocispec.Descriptor{}, "", "", errors.New("missing user input reference")
+	}
+	var tagOrDigestRef string
+	var fullRef string
+	var printOut string
+	switch inputType {
+	case inputTypeRegistry:
+		ref, err := registry.ParseReference(reference)
+		if err != nil {
+			return ocispec.Descriptor{}, "", "", fmt.Errorf("failed to resolve user input reference: %w", err)
+		}
+		tagOrDigestRef = ref.Reference
+		fullRef = ref.Registry + "/" + ref.Repository
+		printOut = ref.Registry + "/" + ref.Repository
+	case inputTypeOCILayout:
+		layoutPath, layoutReference, err := parseOCILayoutReference(reference)
+		if err != nil {
+			return ocispec.Descriptor{}, "", "", fmt.Errorf("failed to resolve user input reference: %w", err)
+		}
+		layoutPathInfo, err := os.Stat(layoutPath)
+		if err != nil {
+			return ocispec.Descriptor{}, "", "", fmt.Errorf("failed to resolve user input reference: %w", err)
+		}
+		if !layoutPathInfo.IsDir() {
+			return ocispec.Descriptor{}, "", "", errors.New("failed to resolve user input reference: input path is not a dir")
+		}
+		tagOrDigestRef = layoutReference
+		printOut = layoutPath
+		fullRef = ociLayout + "/" + filepath.Base(layoutPath)
```

Comment:
I just noticed `ArtifactReference` is never used in `Signer.Sign()` but in `notation.Sign()`. Can you recall why we have `ArtifactReference` in `SignerSignOptions` in the first place?

Reply:
Also, `ArtifactReference` does not need to be a full reference for `Sign()`.

Reply:
Do you think we need to update `notation-go` accordingly?

Reply:
SGTM